**Project Description:**
Predict if a customer is happy or not based on the answers they give to questions asked.

A customer's happiness is recorded in the 'Y' column of the ACME-HappinessSurvey2020.csv file. X values X1 to X6 record the answers to each question in the survey by the customers. Hence these X values are the features we will use to predict whether the custome is happy or not.

Below is the prediction model created for this purpose.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import model_selection
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn import metrics

df = pd.read_csv('ACME-HappinessSurvey2020.csv')

# split data into x and y
y = df.iloc[:,0]
x = df.iloc[:,1:7]

# split into train and test
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.4,random_state=100)

# generate the model
model = MLPClassifier(solver='adam', alpha=1e-5,
                     hidden_layer_sizes=(8), random_state=3)
# train
model.fit(x,y)

# predict
y_pred=model.predict(X_test)

# evaluate performance
print("Accuracy: ",metrics.accuracy_score(y_test, y_pred))

Accuracy:  0.7450980392156863


From the data we have taken the first column 'Y' as the y data that we are going to create a model to predict. The 'x' data that we will use to predict the y value consists of xolumns X1 to X6.

Then we split the data with train_test_split to define which rows we are going to use to train the model and which rows we are going to use to test the trained model on. The test size of '0.4' indicates that we are going to be testing with 40% of the original dataset. Therefore the training data consists of 60% of the original dataset.

Then we create a an MLPClaissifier neural network model using the solver 'adam' and a hidden layer size of 8. After training the model and predicting on the test dataset, a prediction accuracy of 74.5% is achieved.

In [ ]:
from sklearn.model_selection import cross_val_score

# Perform K-fold cross-validation with 5 folds
num_folds = 5
scores = cross_val_score(model, x, y, cv=num_folds)

# Print the cross-validation scores
print("Cross-Validation Scores:", scores)

# Calculate and print the mean accuracy
mean_accuracy = np.mean(scores)
print("Mean Accuracy:", mean_accuracy)

Cross-Validation Scores: [0.61538462 0.6        0.72       0.64       0.52      ]
Mean Accuracy: 0.6190769230769231


Since our dataset is rather small and splitting such a small dataset can result in potentially unreliable and inconsistent results, we will use K-Fold Cross Validation in an attempt to get a more reliable accuracy score based on the dataset. This ensures that the randomly chosen training and testing data does not so easily skew the results since in K-Fold Cross Validation the data is split K times such that in one split we are able to test on one set of data and in another split we are able to train on that same data.

From the cross validation scores we can see a mean accuracy of about 62%. This has dropped from the previous accuracy score since the first split happened to test on a split of data that performed well, while this displays a more accurate result of training with such a small dataset with not many direct correlations between x and y to be found.

**Conclusion:**
The model can perform well but only when tested against selected parts of the dataset. This is because the dataset is not sufficiently large enough to perform either adequate training or adequate testing. With a larger dataset the model could be further trained to find deeper correlations in the data and testing could retrieve more reliable results that are more indicative of the model's performance with future data.